### ISSUE TITLE
Unresolved reference to symbol 

### ISSUE THREAD

**Issue by jaelys** [oleg-shilo/wixsharp#360]
```
project.ResolveWildCards().FindFile((f) => f.Name.EndsWith("HiPace.exe")).First().Associations = new[] { new FileAssociation("hps", "application/hps", "open", "-s \"%1\"") { Icon = "hps.ico" } };
```

```
new InstalledFileAction("DPInstx86.exe", "/S /LM", Return.asyncNoWait, When.Before, Step.InstallFinalize, Condition.NOT_Installed, Sequence.InstallExecuteSequence) { Execute = Execute.deferred, Impersonate = false, Condition = new Condition("NOT VersionNT64 AND VersionNT > 500") },
new InstalledFileAction("HiDaemon.exe", "", Return.asyncNoWait, When.Before, Step.InstallFinalize, Condition.NOT_Installed, Sequence.InstallExecuteSequence) { Execute = Execute.commit }
```
Unresolved reference to symbol 'File:hps.ico' in section 'Product:
Unresolved reference to symbol 'File:DPInstx86.exe' in section 'Product:
Unresolved reference to symbol 'File:HiDaemon.exe' in section 'Product:

1.4.10  build ok
1.6.4.3 build error

Can help me ?

**Comment by oleg-shilo** [comment]
Can you please provide a simple project with a complete source code that shows the problem. Without it it's impossible to understand what `ResolveWildCards` yielded at runtime.

**Comment by jaelys** [comment]
**Program.cs**
```
{
            var project = new Project(productInfo.Name,
                new LaunchCondition("VersionNT >=\"601\" OR VersionNT64 >=\"601\"", "Don't supperor Windows Vista or XP."),
                new Dir(@"%ProgramFiles%\HABOOK\" + productInfo.Name,
                    new Files(@"%SourceDir%\*")),
                new Dir(@"%ProgramMenu%\" + productInfo.Name,
                    new ExeFileShortcut("HiTeach", "[INSTALLDIR]HiTeach.exe", ""),
                    new ExeFileShortcut("HiTeach SelfPace", @"[INSTALLDIR]HiTeachSelfPace.exe", "")),
                new Dir(@"%Desktop%", new ExeFileShortcut("HiTeach", "[INSTALLDIR]HiTeach.exe", "")),
                new SetPropertyAction("ARPINSTALLLOCATION", "[INSTALLDIR]", Return.check, When.After, Step.CostFinalize, Condition.NOT_Installed), //重要：必加才能透過MSI API屬性InstallLocation取得路徑
                new InstalledFileAction("DPInstx86.exe", "/S /LM", Return.asyncNoWait, When.Before, Step.InstallFinalize, Condition.NOT_Installed, Sequence.InstallExecuteSequence) { Execute = Execute.deferred, Impersonate = false, Condition = new Condition("NOT VersionNT64 AND VersionNT > 500") },
                new InstalledFileAction("DPInstx64.exe", "/S /LM", Return.asyncNoWait, When.Before, Step.InstallFinalize, Condition.NOT_Installed, Sequence.InstallExecuteSequence) { Execute = Execute.deferred, Impersonate = false, Condition = new Condition("VersionNT64 AND VersionNT > 500") },
                new InstalledFileAction("HiTeachDaemon.exe", "", Return.asyncNoWait, When.Before, Step.InstallFinalize, Condition.NOT_Installed, Sequence.InstallExecuteSequence) { Execute = Execute.commit })
            {
                InstallerVersion = 405,
                Platform = Platform.x86,
                InstallScope = InstallScope.perMachine,
                UpgradeCode = upgradeCode,
                ProductId = productId,
                Version = version,
                ControlPanelInfo = productInfo,
                OutFileName = productInfo.Name
            };
            project.ResolveWildCards().FindFile((f) => f.Name.EndsWith("HiTeach.exe")).First().Associations = new[] { new FileAssociation("hte", "application/hte", "open", "-s \"%1\"") { Icon = "hte.ico" }, new FileAssociation("knw") { Command = "preview", Icon = "knw.ico" } };
            project.ResolveWildCards().FindFile((f) => f.Name.EndsWith("HiTeachSelfPace.exe")).First().Associations = new[] { new FileAssociation("hts", "application/hts", "open", "-s \"%1\"") { Icon = "hts.ico" } };
            project.LightOptions += "-sice:ICE30";
            project.Encoding = Encoding.UTF8;         
            project.SetNetFxPrerequisite("WIX_IS_NETFRAMEWORK_461_OR_LATER_INSTALLED >= '#394254'", "Requires .NET Framework 4.6.1 or higher.");           
            project.MajorUpgradeStrategy = new MajorUpgradeStrategy
            {
                RemoveExistingProductAfter = Step.InstallValidate,
                UpgradeVersions = VersionRange.OlderThanThis,
                PreventDowngradingVersions = VersionRange.NewerThanThis,
                NewerProductInstalledErrorMessage = "Newer version already installed !!"
            };           
            project.WixSourceGenerated += Compiler_WixSourceGenerated;
            project.ControlPanelInfo.NoModify = true;
            project.ControlPanelInfo.NoRepair = true;
            project.UI = WUI.WixUI_ProgressOnly;          
            var msipath = project.BuildMsi();
            return msipath;          
        }
```

**Product.wxs**

```
<?xml version="1.0" encoding="utf-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi" xmlns:netfx="http://schemas.microsoft.com/wix/NetFxExtension">
  <Product Id="16c16919-cc68-4c65-83dd-d8b03b461b0b" Name="HiTeach" Language="1033" Codepage="Windows-1252" Version="3.0.0.0" UpgradeCode="f164b0c3-f665-49c2-bea0-39f5535ff78e" Manufacturer="HABOOK Group">
    <Package InstallerVersion="405" Compressed="yes" Platform="x86" SummaryCodepage="Windows-1252" Languages="1033" InstallScope="perMachine" />
    <Media Id="1" Cabinet="HiTeach_.cab" EmbedCab="yes" />
    <Condition Message="Don't supperor Windows Vista or XP."><![CDATA[VersionNT >="601" OR VersionNT64 >="601"]]></Condition>
    <Condition Message="Requires .NET Framework 4.6.1 or higher."><![CDATA[WIX_IS_NETFRAMEWORK_461_OR_LATER_INSTALLED >= "#394254"]]></Condition>
    <Directory Id="TARGETDIR" Name="SourceDir">
      <Directory Id="ProgramFilesFolder" Name="ProgramFilesFolder">
        <Directory Id="ProgramFilesFolder.HABOOK" Name="HABOOK">
          <Directory Id="INSTALLDIR" Name="HiTeach">
            <Component Id="Component.Hi.ico_1005481004" Guid="d9fecb59-1240-4e0c-8a9b-96c067ad41ee">
              <File Id="Hi.ico_1005481004" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\Hi.ico" />
            </Component>
            <Component Id="Component.HiTeach.exe_1486646089" Guid="d9fecb59-1240-4e0c-8a9b-96c010ea8302">
              <File Id="HiTeach.exe_1486646089" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\HiTeach.exe" />
              <ProgId Id="hte.file" Advertise="no" Description="hte file" Icon="hte.ico" IconIndex="0">
                <Extension Id="hte" ContentType="application/hte">
                  <Verb TargetFile="HiTeach.exe_1486646089" Id="open" Command="open" Argument="-s &quot;%1&quot;" />
                </Extension>
              </ProgId>
              <ProgId Id="knw.file" Advertise="no" Description="knw file" Icon="knw.ico" IconIndex="0">
                <Extension Id="knw" ContentType="application/knw">
                  <Verb TargetFile="HiTeach.exe_1486646089" Id="preview" Command="preview" Argument="&quot;%1&quot;" />
                </Extension>
              </ProgId>
            </Component>
            <Component Id="Component.HiTeachDaemon.exe_295668060" Guid="d9fecb59-1240-4e0c-8a9b-96c0d888ae96">
              <File Id="HiTeachDaemon.exe_295668060" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\HiTeachDaemon.exe" />
            </Component>
            <Component Id="Component.HiTeachSelfPace.exe_950270141" Guid="d9fecb59-1240-4e0c-8a9b-96c0f5759a8b">
              <File Id="HiTeachSelfPace.exe_950270141" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\HiTeachSelfPace.exe" />
              <ProgId Id="hts.file" Advertise="no" Description="hts file" Icon="hts.ico" IconIndex="0">
                <Extension Id="hts" ContentType="application/hts">
                  <Verb TargetFile="HiTeachSelfPace.exe_950270141" Id="open" Command="open" Argument="-s &quot;%1&quot;" />
                </Extension>
              </ProgId>
            </Component>
            <Component Id="Component.hte.ico_873214712" Guid="d9fecb59-1240-4e0c-8a9b-96c0fb054b62">
              <File Id="hte.ico_873214712" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\hte.ico" />
            </Component>
            <Component Id="Component.hts.ico_1706376626" Guid="d9fecb59-1240-4e0c-8a9b-96c07b2fcb0d">
              <File Id="hts.ico_1706376626" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\hts.ico" />
            </Component>
            <Component Id="Component.knw.ico_1702968835" Guid="d9fecb59-1240-4e0c-8a9b-96c0c83ea588">
              <File Id="knw.ico_1702968835" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\knw.ico" />
            </Component>
            <Directory Id="INSTALLDIR.Driver" Name="Driver">
              <Directory Id="INSTALLDIR.Driver.ftdi" Name="ftdi">
                <Component Id="Component.DPInstx64.exe_825070930" Guid="d9fecb59-1240-4e0c-8a9b-96c04593347d">
                  <File Id="DPInstx64.exe_825070930" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\Driver\ftdi\DPInstx64.exe" />
                </Component>
                <Component Id="Component.DPInstx86.exe_122918288" Guid="d9fecb59-1240-4e0c-8a9b-96c07bd6dcb5">
                  <File Id="DPInstx86.exe_122918288" Source="..\..\..\HiTeach_MSI_test-Source\ProgramFile\Driver\ftdi\DPInstx86.exe" />
                </Component>
              </Directory>
            </Directory>
          </Directory>
        </Directory>
      </Directory>
      <Directory Id="ProgramMenuFolder" Name="ProgramMenuFolder">
        <Directory Id="ProgramMenuFolder.HiTeach" Name="HiTeach">
          <Component Id="HiTeach___" Guid="d9fecb59-1240-4e0c-8a9b-96c0e545572b">
            <Shortcut Id="ProgramMenuFolder.HiTeach.HiTeach___" WorkingDirectory="INSTALLDIR" Target="[INSTALLDIR]HiTeach.exe" Name="HiTeach.lnk" />
            <RemoveFolder Id="ProgramMenuFolder.HiTeach" On="uninstall" />
            <RegistryKey Root="HKCU" Key="Software\WixSharp\Used">
              <RegistryValue Value="0" Type="string" KeyPath="yes" />
            </RegistryKey>
          </Component>
          <Component Id="HiTeach_SelfPace" Guid="d9fecb59-1240-4e0c-8a9b-96c021a1110e">
            <Shortcut Id="ProgramMenuFolder.HiTeach.HiTeach_SelfPace" WorkingDirectory="INSTALLDIR" Target="[INSTALLDIR]HiTeachSelfPace.exe" Name="HiTeach SelfPace.lnk" />
            <RegistryKey Root="HKCU" Key="Software\WixSharp\Used">
              <RegistryValue Value="0" Type="string" KeyPath="yes" />
            </RegistryKey>
          </Component>
        </Directory>
      </Directory>
      <Directory Id="DesktopFolder" Name="DesktopFolder">
        <Component Id="HiTeach.1" Guid="d9fecb59-1240-4e0c-8a9b-96c0ba2357bd">
          <Shortcut Id="DesktopFolder.HiTeach.1" WorkingDirectory="INSTALLDIR" Target="[INSTALLDIR]HiTeach.exe" Name="HiTeach.lnk" />
          <RegistryKey Root="HKCU" Key="Software\WixSharp\Used">
            <RegistryValue Value="0" Type="string" KeyPath="yes" />
          </RegistryKey>
        </Component>
      </Directory>
    </Directory>
    <PropertyRef Id="WIX_IS_NETFRAMEWORK_461_OR_LATER_INSTALLED" />
    <UI />
    <Upgrade Id="f164b0c3-f665-49c2-bea0-39f5535ff78e">
      <UpgradeVersion Minimum="0.0.0.0" Maximum="3.0.0.0" IncludeMinimum="yes" IncludeMaximum="no" Property="UPGRADEFOUND" />
      <UpgradeVersion Minimum="3.0.0.0" IncludeMinimum="no" Property="NEWPRODUCTFOUND" />
    </Upgrade>
    <Icon Id="app_icon.ico" SourceFile="..\..\..\HiTeach_MSI-Source\ProgramFile\Hi.ico" />
    <CustomAction Id="SetProp_ARPINSTALLLOCATION" Property="ARPINSTALLLOCATION" Value="[INSTALLDIR]" Return="check" Execute="immediate" />
    <CustomAction Id="DPInstx86.exe" ExeCommand="/S /LM" FileKey="DPInstx86.exe" Return="asyncNoWait" Impersonate="no" Execute="deferred" />
    <CustomAction Id="DPInstx64.exe" ExeCommand="/S /LM" FileKey="DPInstx64.exe" Return="asyncNoWait" Impersonate="no" Execute="deferred" />
    <CustomAction Id="HiTeachDaemon.exe" ExeCommand="" FileKey="HiTeachDaemon.exe" Return="asyncNoWait" Execute="commit" />
    <CustomAction Id="Set_ARPNOMODIFY" Property="ARPNOMODIFY" Value="True" Return="check" Execute="immediate" />
    <CustomAction Id="PreventDowngrading" Error="Newer version already installed !!" />
    <Property Id="ARPNOREPAIR" Value="True" />
    <Property Id="ARPPRODUCTICON" Value="app_icon.ico" />
    <Feature Id="Complete" Title="Complete" Absent="allow" Level="1">
      <ComponentRef Id="Component.Hi.ico_1005481004" />
      <ComponentRef Id="Component.HiTeach.exe_1486646089" />
      <ComponentRef Id="Component.HiTeachDaemon.exe_295668060" />
      <ComponentRef Id="Component.HiTeachSelfPace.exe_950270141" />
      <ComponentRef Id="Component.hte.ico_873214712" />
      <ComponentRef Id="Component.hts.ico_1706376626" />
      <ComponentRef Id="Component.knw.ico_1702968835" />
      <ComponentRef Id="Component.DPInstx64.exe_825070930" />
      <ComponentRef Id="Component.DPInstx86.exe_122918288" />
      <ComponentRef Id="HiTeach___" />
      <ComponentRef Id="HiTeach_SelfPace" />
      <ComponentRef Id="HiTeach.1" />
    </Feature>
    <InstallExecuteSequence>
      <Custom Action="SetProp_ARPINSTALLLOCATION" After="CostFinalize"> (NOT Installed) </Custom>
      <Custom Action="DPInstx86.exe" Before="InstallFinalize"><![CDATA[NOT VersionNT64 AND VersionNT > 500]]></Custom>
      <Custom Action="DPInstx64.exe" Before="InstallFinalize"><![CDATA[VersionNT64 AND VersionNT > 500]]></Custom>
      <Custom Action="HiTeachDaemon.exe" Before="InstallFinalize"> (NOT Installed) </Custom>
      <Custom Action="Set_ARPNOMODIFY" After="CostFinalize"> (NOT Installed) </Custom>
      <Custom Action="PreventDowngrading" After="FindRelatedProducts">NEWPRODUCTFOUND</Custom>
      <RemoveExistingProducts After="InstallValidate" />
    </InstallExecuteSequence>
    <InstallUISequence>
      <Custom Action="PreventDowngrading" After="FindRelatedProducts">NEWPRODUCTFOUND</Custom>
    </InstallUISequence>
  </Product>
</Wix>
```

**error**
```
Unresolved reference to symbol 'File:hte.ico' in section 'Product:{xxxxxxxx ...
Unresolved reference to symbol 'File:knw.ico' in section 'Product:{xxxxxxxx ...
Unresolved reference to symbol 'File:hts.ico' in section 'Product:{xxxxxxxx ...
Unresolved reference to symbol 'File:DPInstx86.exe' in section 'Product:{xxxxxxxx ...
Unresolved reference to symbol 'File:DPInstx64.exe' in section 'Product:{xxxxxxxx ...
Unresolved reference to symbol 'File:HiTeachDaemon.exe' in section 'Product:{xxxxxxxx ...
```

**build error in 1.6 ver up
build success in 1.5 ver down**

**Comment by jaelys** [comment]
Dear @oleg-shilo 
I have updated the content of the question, please guide you thanks

**Comment by jaelys** [comment]
Dear @oleg-shilo

Thank you for your guidance, ICON can run normally in the new version, but I still can not understand how to amend InstalledFileAction?

**Comment by jaelys** [comment]
Dear @oleg-shilo

Keeping the source directory as a universal character is the most convenient for me, otherwise I may have to deal with very many files and directories separately, and I will try to use #2 to comply with the new version WixSharp

About
`Compiler.AutoGeneration.LegacyDefaultIdAlgorithm = true;`
I would like to try to keep up with the new practice, if not, I will use this solution

Thank you very much for your guidance and let me benefit

**Comment by oleg-shilo** [comment]
Not a problem. That should work for you. You'll need to discover the id of the "DPInstx86.exe" and then ensure `InstallFileAction` has it as a key.

**Comment by jaelys** [comment]
Dear @oleg-shilo

I have a problem with the new version use
`Compiler.AutoGeneration.LegacyDefaultIdAlgorithm = true;`
Will cause
```
project.ResolveWildCards().FindFile((f) =>
f.Name.EndsWith("HiTeach.exe") ||
f.Name.EndsWith("HiTeachDCServer.exe")
).ForEach((x) =>
    {
        x.FirewallExceptions = new[] { new FirewallException("HABOOK." + Path.GetFileNameWithoutExtension(x.Name)) { Scope = FirewallExceptionScope.any, Profile = FirewallExceptionProfile.all, Protocol = FirewallExceptionProtocol.tcp, IgnoreFailure = true } };
    });
```
no work
Wxs does not appear corresponding
`<FirewallException>`

**Comment by oleg-shilo** [comment]
You will need to provide more details. It's not clear what is the actual problem. The tanslation tool that you used did not produced a meaningful output.

Please have a look at "FirewallException" sample in the samples library.

**Comment by jaelys** [comment]
Dear @oleg-shilo ，
I'm sorry, I uploaded a simple example, thank you for your assistance
[sample.zip](https://github.com/oleg-shilo/wixsharp/files/2186922/sample.zip)

HiTeach_wixsharp1.5.wxs (have FirewallException)
HiTeach_wixsharp1.8.wxs (No FirewallException)

**Comment by oleg-shilo** [comment]
txs for the sample, but have had a look at "FirewallException" sample in the samples library?

**Comment by oleg-shilo** [comment]
Sorry but your sample cannot be build:
![image](https://user-images.githubusercontent.com/16729806/42699492-487e9f88-8704-11e8-8098-3f6d39d0c0e8.png)

You'll need to prepare some simpler test solution of a "Hello World" style.

**Comment by jaelys** [comment]
Dear @oleg-shilo ，
My fault
[sample.zip](https://github.com/oleg-shilo/wixsharp/files/2196345/sample.zip)
Thank you

**Comment by oleg-shilo** [comment]
Sorry but the last sample project exhibits the very same problem:
![image](https://user-images.githubusercontent.com/16729806/42756761-f0022066-8940-11e8-8895-384c1c46c83c.png)

Your sample is still full of the hard-codded absolute paths. I suggest you test it first on your PC. Copy it in a new location and see if you can load and build it from there. 

Another point, I see that you are calling `ResolveWildCards` multiple times, while you expected to do it only once:
```C#
project.ResolveWildCards();

WixSharp.File hiTech = project.FindFile(f => f.Name.EndsWith("HiTeach.exe")).First();

hiTech.AddAssociation(new FileAssociation("hte", "application/hte", "open", "-s \"%1\"")
                        {
                            Icon = "hte.ico"
                        });

hiTech.AddAssociation(new FileAssociation("knw")
                        {
                            Command = "preview",
                            Icon = "knw.ico"
                        } );

project.FindFile(f => f.Name.EndsWith("HiTeach.exe") || f.Name.EndsWith("HiTeachDaemon.exe"))
        .ForEach(x =>
                {
                    x.FirewallExceptions = new[] 
                    {
                        new FirewallException("HABOOK." + Path.GetFileNameWithoutExtension(x.Name))
                        {
                            Scope = FirewallExceptionScope.any,
                            Profile = FirewallExceptionProfile.all,
                            Protocol = FirewallExceptionProtocol.tcp,
                            IgnoreFailure = true
                        }
                    };
                });
```

**Comment by jaelys** [comment]
Dear @oleg-shilo,

[Sample2.zip](https://github.com/oleg-shilo/wixsharp/files/2199830/Sample2.zip)

I am sorry, wasting your time, I have fixed hard-codded absolute paths
WixSharp 1.5 and 1.8 switching is a bit broken, fixed, should be normal
There are two files in the Sample folder.
```
HiTeach_wixsharp1.5.wxs
HiTeach_wixsharp1.8.wxs
```
Is the difference between the two versions, 1.8 has no <FirewallException>

Thank you for taking the time to be grateful. Thank you very much.

**Comment by oleg-shilo** [comment]
I just tested your latest solution and it can build now. Txs.

Though int the generated WXS the `FireWallException` element is where it is expected to be:

![image](https://user-images.githubusercontent.com/16729806/42806630-3650a87e-89f2-11e8-8f76-406e9b8fe7dd.png)

**Comment by jaelys** [comment]
Dear @oleg-shilo,
Sample build in  1.5.0 lib 
Can you try to update to wixsharp 1.8.0 on nuget
You should see HiTeach.wxs without FireWallException element ^^

**Comment by oleg-shilo** [comment]
Great, finally we got on the same page.

I was able to reproduce the problem. Indeed it is caused by the recent change in the v1.8. The problem will be corrected in the very next release.

Until the fix is available please use the following work around (marked with `// the solution:`).
Please note I have slightly reformatted your solution as it was very difficult to read.

```c#
static string BuildMsi(ProductInfo productInfo, Guid upgradeCode, Guid productId, Version version)
{
    var project = new Project(productInfo.Name,
        new LaunchCondition("VersionNT >=\"601\" OR VersionNT64 >=\"601\"", "Don't supperor Windows Vista or XP."),
        new Dir(@"%ProgramFiles%\HABOOK\" + productInfo.Name,
            new Files(@"Source\*.*")),
        new Dir(@"%ProgramMenu%\" + productInfo.Name,
            new ExeFileShortcut("HiTeach", "[INSTALLDIR]HiTeach.exe", "")),
        new Dir(@"%Desktop%",
            new ExeFileShortcut("HiTeach", "[INSTALLDIR]HiTeach.exe", "")),
        new SetPropertyAction("ARPINSTALLLOCATION", "[INSTALLDIR]", Return.check, When.After, Step.CostFinalize, Condition.NOT_Installed) //重要：必加才能透過MSI API屬性InstallLocation取得路徑                                                                                                                                              
        )
    {
        InstallerVersion = 405,
        Platform = Platform.x86,
        InstallScope = InstallScope.perMachine,
        UpgradeCode = upgradeCode,
        ProductId = productId,
        Version = version,
        ControlPanelInfo = productInfo,
        OutFileName = productInfo.Name
    };

    project.ResolveWildCards();
    project.FindFile((f) => f.Name.EndsWith("HiTeach.exe"))
           .FirstOrDefault()
           .Associations = new[]
           {
               new FileAssociation("hte", "application/hte", "open", "-s \"%1\"")
               {
                   Icon = "hte.ico"
               },
               new FileAssociation("knw")
               {
                   Command = "preview",
                   Icon = "knw.ico"
               }
           };

    project.FindFile((f) => f.Name.EndsWith("HiTeach.exe") || f.Name.EndsWith("HiTeachDaemon.exe"))
           .ForEach((x) =>
    {
        Console.WriteLine(x.Name);
       
        // the solution: replace 'x.FirewallExceptions' with 'x.GenericItems'
        x.GenericItems = new[]
        {
            new FirewallException("HABOOK." + Path.GetFileNameWithoutExtension(x.Name))
            {
                Scope = FirewallExceptionScope.any,
                Profile = FirewallExceptionProfile.all,
                Protocol = FirewallExceptionProtocol.tcp,
                IgnoreFailure = true
            }
        };
    });
```

**Comment by jaelys** [comment]
Dear @oleg-shilo 

Thank you very much for your patience.
I will pay more attention to the quality of the article and code in the future.
Thanks again for this great project.

Sorry my English is from Google Translate

**Comment by oleg-shilo** [comment]
No worries,

> Sorry my English is from Google Translate

Yes Google confused us a bit. But... we ended up OK after all :)
